torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h

```diff
@@ -0,0 +1,256 @@
+  }
+
+ private:
+  std::vector<int64_t> output_shape_;
+  std::vector<int64_t> broadcast_dims_;
+};
+
+template <class OutType, class ArgType>
+struct CastOpRecord : RecordFunctor {
+  CastOpRecord(
+      std::vector<size_t> _args,
+      std::vector<size_t> _outputs,
+      std::function<OutType(NvfDataType, ArgType)> fusion_op,
+      NvfDataType dtype)
+      : RecordFunctor(std::move(_args), std::move(_outputs)),
+        fusion_op_(fusion_op),
+        dtype_(dtype) {}
+
+  void operator()(FusionDefinition& fd) final {
+    auto arg = dynamic_cast<ArgType>(fd.fusion_state.at(args.at(0)));
+    auto output = fusion_op_(dtype_, arg);
+    fd.fusion_state.at(outputs.at(0)) = output;
+  }
+
+ private:
+  std::function<OutType(NvfDataType, ArgType)> fusion_op_;
+  NvfDataType dtype_;
+};
+
+template <class OutType, class... ArgTypes>
+struct OpRecord : RecordFunctor {
+  OpRecord(
+      std::vector<size_t> _args,
+      std::vector<size_t> _outputs,
+      std::function<OutType(ArgTypes...)> fusion_op)
+      : RecordFunctor(std::move(_args), std::move(_outputs)),
+        fusion_op_(fusion_op) {}
+
+  template <class TupleType, std::size_t... Is>
+  OutType opFunc(
+      FusionDefinition& fd,
+      TupleType& tp,
+      std::index_sequence<Is...>) {
+    return fusion_op_(
+        dynamic_cast<typename std::tuple_element<Is, TupleType>::type>(
+            fd.fusion_state.at(args.at(Is)))...);
+  }
+
+  void operator()(FusionDefinition& fd) final {
+    using arg_tuple_t = std::tuple<ArgTypes...>;
+    auto indices =
+        std::make_index_sequence<std::tuple_size<arg_tuple_t>::value>();
+    arg_tuple_t inputs;
+    auto output = opFunc(fd, inputs, indices);
+    fd.fusion_state.at(outputs.at(0)) = output;
+  }
+
+ private:
+  std::function<OutType(ArgTypes...)> fusion_op_;
+};
+
+struct ReductionOpRecord : RecordFunctor {
+  ReductionOpRecord(
+      std::vector<size_t> _args,
+      std::vector<size_t> _outputs,
+      std::function<
+          NvfTensorView*(NvfTensorView*, std::vector<int>&, bool, NvfDataType)>
+          fusion_op,
+      std::vector<int> axes,
+      bool keep_dim,
+      NvfDataType dtype)
+      : RecordFunctor(std::move(_args), std::move(_outputs)),
+        fusion_op_(fusion_op),
+        axes_(std::move(axes)),
+        keep_dim_(keep_dim),
+        dtype_(dtype) {}
+
+  void operator()(FusionDefinition& fd) final {
+    auto arg = fd.fusion_state.at(args.at(0))->as<NvfTensorView>();
+    auto output = fusion_op_(arg, axes_, keep_dim_, dtype_);
+    fd.fusion_state.at(outputs.at(0)) = output;
+  }
+
+ private:
+  std::function<
+      NvfTensorView*(NvfTensorView*, std::vector<int>&, bool, NvfDataType)>
+      fusion_op_;
+  std::vector<int> axes_;
+  bool keep_dim_;
+  NvfDataType dtype_;
+};
+
+struct VarianceOpRecord : RecordFunctor {
+  VarianceOpRecord(
+      std::vector<size_t> _args,
+      std::vector<size_t> _outputs,
+      std::vector<int>& axes,
+      int64_t correction,
```

Comment:
Should correction be a `bool` argument?

Reply:
I was following this line in our composite variance op. 
 Is it not correct? https://github.com/pytorch/pytorch/blob/18e2aeae01dd8fb7b12d118828542f1321a242b6/torch/csrc/jit/codegen/cuda/ops/normalization.h#L52-L56.

Reply:
The python frontend takes a bool but aten uses an integer. I guess we ended up with both definitions in the composite ops.